File: AlgorithmsPackJavaSolutions/Longest Palindromic Substring.java
public class Solution {
    private int[][] memory;
    private String str;
    
    public String longestPalindrome(String s) {
        str = s;
        int n = s.length();
        memory = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                memory[i][j] = -1;
            }
        }
        
        int length = 0;
        int startIndex = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int len = j - i + 1;
                if (isValid(i, j) && len > length) {
                    length = len;
                    startIndex = i;
                }
            }
        }
        return s.substring(startIndex, startIndex + length);
    }
    
    private boolean isValid(int i, int j) {
        if (i >= j) {
            return true;
        }
        
        if (memory[i][j] != -1) {
            return memory[i][j] == 1;
        }
        
        if (str.charAt(i) == str.charAt(j)) {
            boolean ret = isValid(i + 1, j - 1);
            memory[i][j] = ret ? 1 : 0;
            return ret;
        }
        
        memory[i][j] = 0;
        return false;
    }
}

File: AlgorithmsPackPythonSolutions/Longest Palindromic Substring.py
class Solution:
    def isValid(self, i: int, j: int, memory: List[List[int]], s: str) -> bool:
        if i >= j:
            return True
        
        if memory[i][j] != -1:
            return memory[i][j]
        
        if s[i] == s[j]:
            ret = self.isValid(i + 1, j - 1, memory, s)
            memory[i][j] = ret
            return ret
        
        memory[i][j] = False
        return False
    
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        memory = [[-1] * n for _ in range(n)]
        length = 0
        startIndex = 0
        
        for i in range(n):
            for j in range(i, n):
                len_ji = j - i + 1
                if self.isValid(i, j, memory, s) and len_ji > length:
                    length = len_ji
                    startIndex = i
        
        return s[startIndex:startIndex + length]
